rust/src/sdk/api.rs

```diff
@@ -0,0 +1,448 @@
+                            panic!("iterating further results in undefined behavior");
+                        }
+                        unsafe {
+                            Some(*self.model.ptr.offset(self.count))
+                        }
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    };
+}
+
+PrimModel! {f32, c_float}
+PrimModel! {f64, c_double}
+PrimModel! {i32, c_int}
+PrimModel! {i64, c_long}
+
+#[derive(Clone, Debug)]
+/// A primitive model of data type `N` cached on the heap. The pointer `PrimitiveModelN` returned
+/// from `get_model_N()` references this memory.
+pub enum PrimitiveModel {
+    F32(Vec<f32>),
+    F64(Vec<f64>),
+    I32(Vec<i32>),
+    I64(Vec<i64>),
+}
+
+/// TODO: this is a mock, replace by sth like the `Client` from #397 or a wrapper around that.
+///
+/// The pointer `PrimitiveModelN` to the cached [`PrimitiveModel`] of primitive data type `N`,
+/// which gets allocated and returned in `get_model_N()`, is valid across the FFI-boundary until
+/// one of the following events happen:
+/// - The [`PrimitiveModel`] memory is freed via a call to [`drop_model()`].
+/// - The [`Client`] memory is freed via a call to [`drop_client()`].
+/// - The model is updated via a call to `update_model_N()`.
+/// - The round ends. (TODO: implement this point when a new round is observed)
+pub struct Client {
+    participant: Participant,
+
+    // counting starts from 1, 0 means not seen yet
+    current_round: u32,
+    checked_round: u32,
+
+    // cached primitive model
+    model: Option<PrimitiveModel>,
+}
+
+#[no_mangle]
+/// Creates a new [`Client`].
+pub extern "C" fn new_client() -> *mut Client {
+    let client = Client {
+        participant: if let Ok(participant) = Participant::new() {
+            participant
+        } else {
+            // TODO: add error handling
+            panic!("participant initialization failed")
+        },
+        current_round: 0,
+        checked_round: 0,
+        model: None,
+    };
+    Box::into_raw(Box::new(client))
+}
+
+#[no_mangle]
+/// Destroys a [`Client`] and frees its allocated memory.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn drop_client(client: *mut Client) {
+    if !client.is_null() {
+        Box::from_raw(client);
+    }
+}
+
+#[no_mangle]
+/// Checks if the next round has started.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn is_next_round(client: *mut Client) -> bool {
+    if client.is_null() {
+        // TODO: add error handling
+        panic!("invalid client");
+    }
+    let client = &mut *client;
```

Comment:
converting a raw pointer to a reference is unsafe. Doesn't rustc complain that this should be within an unsafe block?

Reply:
it actually is inside an unsafe block, because the whole function is declared unsafe.

Reply:
I see. I'd still put it within an `unsafe` block to make it more obvious which part of the function is actually unsafe. It makes auditing the code easier.

Reply:
i was thinking about this as well, i definitely see the benefit of a pinpointing `unsafe` block. but in case of multiple `unsafe` blocks per function, this might become harder to read, hence i chose the `unsafe fn`. also, `unsafe fn` makes this very clear to the caller, otherwise it would look like a safe `fn` on first glance.

if we use both `unsafe fn` and `unsafe` blocks at the same time, we get a redundancy warning:
```
unnecessary `unsafe` block
`#[warn(unused_unsafe)]` on by defaultrustc(unused_unsafe)
api.rs: because it's nested under this `unsafe` fn
```
we might just disable the warning here for the sake of having `unsafe` blocks as well. what do you think?

Reply:
I'm not sure actually because I have very little experience with unsafe. We should ask the community about this. The rust-users forum seems like a good place to get feedback.

Reply:
This seems relevant: https://github.com/rust-lang/rust/issues/69173

Reply:
thanks for the link. i would suggest that we then use `unsafe fn` in combination with explicit `unsafe` blocks. also, we shouldn't disable the warning and just keep it emitting until the lint change gets merged.